### PR TITLE
Add string slicing extension methods.

### PR DESCRIPTION
**Pull request by AtsushiKan**
https://github.com/dotnet/corefx/issues/24072

### CODE REVIEW

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs [hunk 1]
```diff
@@ -70,19 +70,93 @@ public static ReadOnlySpan<char> AsReadOnlySpan(this string text)
...
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index or <paramref name="length"/> is not in range.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> AsReadOnlySpan(this string text, int start, int length)
+        {
+            if (text == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
+            if ((uint)start > (uint)text.Length || (uint)length > (uint)(text.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
```
Comment (stephentoub) [inline src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Couldn't it be length not start that's out of range?

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs [hunk 2]
```diff
@@ -70,19 +70,93 @@ public static ReadOnlySpan<char> AsReadOnlySpan(this string text)
...
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="text"/> is a null reference (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index or <paramref name="length"/> is not in range.
+        /// </exception>
+        public static ReadOnlyMemory<char> AsReadOnlyMemory(this string text, int start, int length)
+        {
+            if (text == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
+            if ((uint)start > (uint)text.Length || (uint)length > (uint)(text.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
```
Comment (stephentoub) [inline src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Ditto

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs (hunk 1 above)
Comment (AtsushiKan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: The portable Span Slice method uses the param name "start" for the combined start/length check - that was the precedent I followed. Personally, I don't have a strong bias towards having a paramname or not on this check - we should decide on one and be consistent about it, though.

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs [hunk 3]
```diff
@@ -70,19 +70,93 @@ public static ReadOnlySpan<char> AsReadOnlySpan(this string text)
...
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index is not in range (&lt;0 or &gt;text.Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> AsReadOnlySpan(this string text, int start)
+        {
+            if (text == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
+            if ((uint)start > (uint)text.Length)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
```
Comment (gfoidl) [inline src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Should this be `if ((uint)start >= (uint)text.Length)`?

Comment (AtsushiKan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: No - `start`==`length` is a valid edge-case input. Just like in the other `Slice` apis and older apis like `String.Substring()`.

**Comment by gfoidl** [comment]
> You get a zero-length range out of it.

You understood. Then it is OK.

PS: sorry for the extra comments. I tried to review (not single comment) then this happened.

**Comment by AtsushiKan** [comment]
@ahsonkhan  - Ping

**Pull request merged**